django/contrib/admin/options.py

```diff
@@ -1583,6 +1584,24 @@ def add_view(self, request, form_url='', extra_context=None):
     def change_view(self, request, object_id, form_url='', extra_context=None):
         return self.changeform_view(request, object_id, form_url, extra_context)
 
+    def get_edited_object_pks(self, request, prefix):
+        pk_pattern = re.compile('{prefix}-\d+-{pk_name}$'.format(prefix=prefix, pk_name=self.model._meta.pk.name))
+        return [
+            value for key, value in request.POST.items() if pk_pattern.match(key)
+        ]
+
+    def get_list_editable_queryset(self, request, prefix):
+        object_pks = self.get_edited_object_pks(request, prefix)
+        queryset = self.get_queryset(request)
+        validate = queryset.model._meta.pk.to_python
+        try:
+            for pk in object_pks:
+                validate(pk)
+        except ValidationError:
+            # Disable optimization
```

Comment:
I don't see the reasoning for disabling the optimization on invalid data. Wouldn't that allow malicious users (unlikely, I guess, if they already have access to the admin) to create memory issues?

Reply:
The reason was that we’d end up with a 500 server error in this case, whereas now we get a validation error. 

The alternative that we could use here is the old approach ‘cl.result_list’, which we know is sensibily limited to just one page.